webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DataStructures.scala

```diff
@@ -579,58 +579,114 @@ final case class BoundPolicyDraft(
    * This is the place where we can check for variable consistency too, since we have the final
    * techniques and expanded variables (like: check that a mandatory variable was expanded to "").
+   *
+   * It's also the place where we add missing variables that are well defined, and where we check for
+   * constrains.
+   *
+   * If there is variable defined here but not in the technique, they are removed.
    */
   def toPolicy(agent: AgentType): Either[String, Policy] = {
     PolicyTechnique.forAgent(technique, agent).flatMap { pt =>
-      val checkedMandatoryValues: Either[Accumulated[RudderError], List[(ComponentId, Variable)]] = expandedVars.accumulatePure {
-        case (cid, variable) =>
-          // check if there is mandatory variable with missing/blank values
-          (variable.spec.constraint.mayBeEmpty, variable.values.exists(v => v == null || v.isBlank)) match {
-            case (false, true) =>
-              // first, if we have a default value, use that
-              variable.spec.constraint.default match {
-                case Some(d) =>
-                  variable
-                    .copyWithSavedValues(variable.values.map(v => if (v.isBlank) d else v))
-                    .map(v => (cid, v))
-                // else it's an error
-                case None    =>
-                  Left(
-                    Inconsistency(
-                      s"Error for policy for directive '${directiveName}' [${id.directiveId.debugString}] in rule '${ruleName}' [${id.ruleId.serialize}]: " +
-                      s"a non optional value is missing for parameter '${variable.spec.description}' [param ID: ${variable.spec.name}]"
-                    )
-                  )
-              }
-            case _             => Right((cid, variable))
-          }
-      }
-
-      checkedMandatoryValues
-        .map(withDefaultVals => {
-          Policy(
-            id,
-            ruleName,
-            directiveName,
-            pt,
-            acceptationDate,
-            NonEmptyList.of(
-              PolicyVars(
-                id,
-                policyMode,
-                withDefaultVals.toMap,
-                originalVars,
-                trackerVariable
-              )
-            ),
-            priority,
-            policyMode,
-            ruleOrder,
-            directiveOrder,
-            overrides
-          )
-        })
-        .left
+      // check that all variables from root section are here and well defined
+      // the lookup for variables is a bit complicated. We have unique variable by name, so for now, we must to only match on that.
+      // Since we use the same technique to build section path, it should be the same (apart in tests) but
+      // I prefer to do the lookup on name
+      val techVarSpecs = technique.getAllVariableSpecs.map { case (cid, s) => (cid.value, (cid, s)) }
+
+      val allVars = expandedVars + (ComponentId(trackerVariable.spec.name, Nil, None) -> trackerVariable)
```

Comment:
tracker variable is not part of the standard ones, but we need to check that it has a value (it's mandatory)